plugins/spike_remove.py

```diff
@@ -0,0 +1,312 @@
+        dlg.ShowModal()
+
+        # The end
+        self._spec_stream = None  # drop reference
+        dlg.Close()
+        self._dlg = None
+        
+        if dlg: # If dlg hasn't been destroyed yet
+            dlg.Destroy()
+
+    def _update_spike_pix(self, npixels, nspikes):
+        """
+        Updates the npixels and nspikes VAs
+        """
+        self.npixels._set_value(npixels, force_write=True)
+        self.nspikes._set_value(nspikes, force_write=True)
+
+    @call_in_wx_main
+    def _update_save_button(self):
+        """
+        Enable the "Save" button iff the data has already been corrected
+        """
+        # Note: it's important that it's run with @call_in_wx_main, to ensure
+        # that it's run asynchronously, and so at dialog creation, it's called
+        # after adding the buttons (which are also asynchronously created).
+        self._dlg.buttons[2].Enable(hasattr(self._spec_stream, "_orig_raw"))
+
+    @call_in_wx_main
+    def save(self, dlg):
+
+        if not hasattr(self._spec_stream, "_orig_raw"):
+            box = wx.MessageDialog(self.main_app.main_frame,
+                   "No correction was applied",
+                   "No correction", wx.OK | wx.ICON_STOP)
+            box.ShowModal()
+            box.Destroy()
+            return
+
+        fn = self.tab_data.acq_fileinfo.value.file_name
+
+        # Store all the data present in the original file => just open it again.
+        das_orig = open_acquisition(fn)
+        das = []
+        for da in das_orig:
+            # Is it the stream that we've corrected?
+            if (self._spec_stream.raw[0].metadata == da.metadata and
+                 self._spec_stream.raw[0].shape == da.shape):
+                das.append(self._spec_stream.raw[0])
+            else:
+                das.append(da)
+
+        # Ask for filename, with default to original filename + _corrected
+        # TODO: smart naming scheme if file already exists.
+        basefn, ext = os.path.splitext(fn)
+        cfn = basefn + "_corrected" + ext
+        cfn = ShowAcquisitionFileDialog(dlg, cfn)
+        exporter = dataio.find_fittest_converter(cfn)
+        if cfn is not None:
+            exporter.export(cfn, das)
+        else:
+            logging.debug("Saving cancelled")
+
+        dlg.Close()
+
+    def correct_data(self, dlg):
+        """
+        Remove the spikes of self._spec_stream
+        """
+        # We keep the original raw data in ._orig_raw, and will update .raw so
+        # that the correction is displayed, but if the user asks to correct again,
+        # we will remove the spike from the raw data, not from the data already
+        # cleaned-up.
+        try:
+            raw_spec_dat = self._spec_stream._orig_raw
+        except AttributeError:  # No orig_raw yet
+            raw_spec_dat = self._spec_stream.raw[0]
+            self._spec_stream._orig_raw = raw_spec_dat
```

Comment:
should `raw_spec_dat` be copied here?

Reply:
It's not needed because the `removespikes_spec()` function doesn't modify the data in-place, it returns a different array. As it's the standard behaviour for functions "loosing" data, I don't think it's needed to put explicit comment about this.